### PR TITLE
Issue 2555

### LINKED ISSUE #2555: Create notifications for events to start
This follows #2461

Create a cronjob for createNotificationsForEventsToStart.

see `apps/backend-functions/src/internals/invitations/events.ts`

### PR DESCRIPTION
**Pull request by fritzschoff**
close #2555 

this function is not tested, cause the firebase emulator is currently not able to handle pub sub functions.

### CODE REVIEW

**Comment by phpgeekfr** [comment]
> close #2555
> 
> this function is not tested, cause the firebase emulator is currently not able to handle pub sub functions.

can't you test if by changing the `.schedule('0 4 * * *')` in main.ts ?

File: apps/backend-functions/src/internals/invitations/events.ts [hunk 1]
```diff
@@ -269,27 +270,23 @@ export async function onInvitationToAnEventUpdate(
...
-  } else if (!!fromOrg) {
-    notification.organization = fromOrg;
-  } else {
-    throw new Error('Did not found invitation sender');
-  }*/
+  const eventsCollection = await db.collection('events').where('date', '<', beginningDateObject).get();
+  const events = eventsCollection.docs.map(doc => doc.data());
+
+  const notifications = events.map(event => createNotification({
+    toUserId: event.toUser.uid,
```
Comment (phpgeekfr) [inline apps/backend-functions/src/internals/invitations/events.ts:284]: `event` does not have `toUser`  . You are mistaking with `invitations`. 
As the comments above says :
```
// 1 Fetch events that are about to start
  // 2 Fetch attendees (invitations accepted)
  // 3 Create notifications
  ```

File: apps/backend-functions/src/internals/invitations/events.ts [hunk 2]
```diff
@@ -269,27 +270,23 @@ export async function onInvitationToAnEventUpdate(
...
+  const beginningDateObject = new Date(beginningDate);
 
-  /*if (!!fromUser) {
-    notification.user = fromUser;
-  } else if (!!fromOrg) {
-    notification.organization = fromOrg;
-  } else {
-    throw new Error('Did not found invitation sender');
-  }*/
+  const eventsCollection = await db.collection('events').where('date', '<', beginningDateObject).get();
```
Comment (phpgeekfr) [inline apps/backend-functions/src/internals/invitations/events.ts:280]: `event` does not have `date` attribute  but `start` and `end`

**Review by phpgeekfr — CHANGES_REQUESTED** [review]

File: apps/backend-functions/src/internals/invitations/events.ts [hunk 3]
```diff
@@ -269,27 +270,23 @@ export async function onInvitationToAnEventUpdate(
...
   // 2 Fetch attendees (invitations accepted)
   // 3 Create notifications
 
-  /*const notification = createNotification({
-    toUserId: toUser.uid,
-    eventId,
-    type: 'eventIsAboutToStart''
-  });*/
+  const beginningDate = Date.now() + 86400000;
+  const beginningDateObject = new Date(beginningDate);
```
Comment (phpgeekfr) [inline apps/backend-functions/src/internals/invitations/events.ts:278]: maybe :
```
const startDate = new Date(Date.now() +( 86400 *1000)); // now plus 1 day
```

**Comment by fritzschoff** [comment]
image
its getting triggered, but I can't see my console.log..

**Comment by fritzschoff** [comment]
George said I could test it in the google cloud console, but there are also no logs.. pretty annoying. Will need to unit test but jest is bugged with the wrong types...

File: apps/backend-functions/src/main.ts [hunk 4]
```diff
@@ -123,7 +123,7 @@ export const twilioWebhook = functions.https.onRequest(_twilioWebhook);
 /**
  * Creates notifications when an event is about to start
  */
-export const scheduledNotifications = functions.pubsub.schedule('0 4 * * *')// every day at 4 AM
+export const scheduledNotifications = functions.pubsub.schedule('*/5 * * * *')// every day at 4 AM
```
Comment (phpgeekfr) [inline apps/backend-functions/src/main.ts:126]: do not commit your  test ;)

Comment (fritzschoff) [reply apps/backend-functions/src/main.ts]: haha, promise I won't

Comment (fritzschoff) [reply apps/backend-functions/src/main.ts]: could be quite expensive

**Comment by phpgeekfr** [comment]
> this function is not tested, cause the firebase emulator is currently not able to handle pub sub functions.

It is now.

File: apps/backend-functions/jest.config.js [hunk 5]
```diff
@@ -12,4 +12,5 @@
   moduleFileExtensions: ['ts', 'js', 'html'],
   setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
   coverageDirectory: '../../coverage/libs/import',
+  testEnvironment: 'node',
 };
```
Comment (xmano) [inline apps/backend-functions/jest.config.js:16]: correct coverage directory to insode backend-functions and keep it last for consistency. please.

File: apps/backend-functions/src/internals/invitations/events.spec.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+import { clearFirestoreData } from "@firebase/testing";
+import { createNotificationsForEventsToStart } from './events';
+
+describe('Events tests', () => {
+
+  beforeAll(async () => {
+    initFunctionsTestMock();
+    loadAdminServices();
+    // To be sure that tests are not polluted
+    await clearFirestoreData({ projectId: getTestingProjectId() });
```
Comment (xmano) [inline apps/backend-functions/src/internals/invitations/events.spec.ts:13]: I think when emulator is started data is not persisted. So it is ok to remove this.

But make sure to clear data in afterAll()

File: apps/backend-functions/src/internals/invitations/events.spec.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+    const users = [{ uid: 'U001' }, { uid: 'U002' }, { uid: 'U003' }];
+
+    const events = [
+      { id: 'E001', start: new Date(Date.now() + oneDay + (oneHour / 2)) }, // Should create a notification
+      { id: 'E002', start: new Date(Date.now() + oneDay + oneDay) }, // Should not create a notification
+      { id: 'E003', start: new Date(Date.now() - oneDay) }, // Should not create a notification
+    ];
+
+    const invitations = [
```
Comment (xmano) [inline apps/backend-functions/src/internals/invitations/events.spec.ts:33]: Can you move this to a fixture file?

This way we can reuse also if needed.

**Review by phpgeekfr — APPROVED** [review]

**Pull request merged**